### PR TITLE
Testbed

### PR DESCRIPTION
**Pull request by nbouchinet-anssi**
Adds a test environment to the ultablue project, it's able to generate a linux distro image which includes the ultrablue server, its dracut module and systemd unit. Once the image is generated one can use it to enroll an ultrablue client and test a remote attested boot.

### CODE REVIEW

File: server/dracut/90ultrablue/module-setup.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: 2022 ANSSI
+# SPDX-License-Identifier: Apache-2.0
+#!/bin/bash
```
Comment (kerneis-anssi) [inline server/dracut/90ultrablue/module-setup.sh]: Does this work? I would expect the shebang to be on the first line.

File: server/unit/ultrablue-server.service [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+Description=ultrablue remote attestation service
+After=bluetooth.service
+Wants=cryptsetup-pre.target
+Before=cryptsetup-pre.target
+DefaultDependencies=no
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/ultrablue-server
+RemainAfterExit=yes
```
Comment (kerneis-anssi) [inline server/unit/ultrablue-server.service]: Out of curiosity, why is this necessary? Is there another service that depends on this one staying active even after the server exits?

File: server/testbed/Makefile [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# SPDX-FileCopyrightText: 2022 ANSSI
+# SPDX-License-Identifier: Apache-2.0
+#
+hostbus := $(shell lsusb | grep Bluetooth | sed -r 's/([^0-9]*([0-9]*)){1}.*/\2/')
```
Comment (kerneis-anssi) [inline server/testbed/Makefile]: Maybe more legible if we use the fact that there is Bus and Device in the output, something like:
`s/Bus ([0-9]+)/\1/`
and likewise with Device for hostaddr.

File: server/testbed/mkosi.default.d/debian/10-mkosi.debian [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# SPDX-FileCopyrightText: 2022 ANSSI
+# SPDX-License-Identifier: Apache-2.0
+
+[Distribution]
+Distribution=debian
+Release=testing
+
+[Packages]
```
Comment (kerneis-anssi) [inline server/testbed/mkosi.default.d/debian/10-mkosi.debian:8]: No clue if there is a concept of "run package" for mkosi, but fyi I had to install `swtpm` to get this command (used in the Makefile) on Debian.

File: README.md [hunk 5]
```diff
@@ -14,18 +14,18 @@ The classical Ultrablue control flow consists of several steps:
 To enroll a phone as a verifier, start the server in enroll mode. This will display a QR code on the terminal. From the phone, run the client app, and tap the **+** icon on the top right corner to show a QR code scanner. On scan, an encrypted Bluetooth Low Energy channel will be established, and the enrollment will run automatically. Upon success, a device card will appear on the home page of the client application.
 
 ### 2. Initramfs configuration (optional)
-The following explains how to integrate Ultrablue in the boot process to guard disk encryption by a secret delivered by the phone.
+Once enrolled, you have to re-generate your initramfs in order to include the **ultrablue dracut module** in it, it can be done using the following dracut command:
 
+```
+dracut --add ultrablue /path/to/initrd --force
```
Comment (kerneis-anssi) [inline README.md:22]: That will only work if the ultrablue module is somehow installed somewhere dracut can find it, right? Maybe we should mention how to do that, or how to use a relative path and point to the module in the git repo.

File: server/testbed/README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Virtual testbed setup
+
+This testbed allows you to generate a linux distribution image (ie. archlinux, fedora, debian) with ultrablue server
+configured in it.
+
+Your machine needs a **TPM2** and **bluetooth** devices.
```
Comment (kerneis-anssi) [inline server/testbed/README.md]: Even with swtpm, it's not a full emulation and you still need a hw one?

File: server/testbed/mkosi.default.d/debian/10-mkosi.debian (hunk 4 above)
Comment (nbouchinet-anssi) [reply server/testbed/mkosi.default.d/debian/10-mkosi.debian:8]: mkosi don't needs swtpm to work, swtpm is used to create a virtual tpm in the makefile before passing it to the mkosi qemu command, this is up to the user to install swtpm.

File: server/unit/ultrablue-server.service (hunk 2 above)
Comment (nbouchinet-anssi) [reply server/unit/ultrablue-server.service]: Are you talking about the RemainAfterExit= option ? I like the idea to keep it active for user informative purpose, it can be dropped without functional drawbacks

File: server/dracut/90ultrablue/module-setup.sh (hunk 1 above)
Comment (nbouchinet-anssi) [reply server/dracut/90ultrablue/module-setup.sh]: Yes it somehow works but its ugly, agree to move it on top of the files

File: server/testbed/Makefile (hunk 3 above)
Comment (nbouchinet-anssi) [reply server/testbed/Makefile]: ACK

File: README.md (hunk 5 above)
Comment (nbouchinet-anssi) [reply README.md:22]: The user have to install the ultrablue dracut module, we could  indeed explain how to do it here

File: server/testbed/README.md (hunk 6 above)
Comment (nbouchinet-anssi) [reply server/testbed/README.md]: My bad, changed it to swtpm requirement

File: server/testbed/Makefile [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	          --tpmstate dir=/tmp/emulated_tpm/ultrablue \
+	          --ctrl type=unixio,path=/tmp/emulated_tpm/ultrablue/swtpm-sock \
+	          --log level=20 --tpm2 --daemon
+	
+	sudo mkosi qemu \
+	    -chardev socket,id=chrtpm,path=/tmp/emulated_tpm/ultrablue/swtpm-sock \
+	    -tpmdev emulator,id=tpm0,chardev=chrtpm \
+	    -device tpm-tis,tpmdev=tpm0 \
+	    -usb -device usb-host,hostbus=${hostbus},hostaddr=${hostaddr}
+
```
Comment (kerneis-anssi) [inline server/testbed/Makefile:41]: Would be nice to add a `make clean` target. Could be in a separate commit.

**Review by kerneis-anssi — APPROVED** [review]
The code looks good to me, with the caveat that the server currently crashes when using a vTPM (see issue #10).

File: server/testbed/Makefile (hunk 7 above)
Comment (nbouchinet-anssi) [reply server/testbed/Makefile:41]: I don't want to add a clean target as it will need to remove file with root privileges
